File: HW/HW11/PythonCode/FloodFill.py
from collections import deque

class FloodFill:
    def floodFill(self, image, sr, sc, color):
        originalColor = image[sr][sc]
        
        if originalColor != color:
            queue = deque([(sr, sc)])
            image[sr][sc] = color

            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            while queue:
                r, c = queue.pop()
                
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < len(image) and 0 <= nc < len(image[0]) and image[nr][nc] == originalColor:
                        image[nr][nc] = color
                        queue.append((nr, nc))
        return image

solution = FloodFill()
image = [[0,1,0],
         [1,1,0],
         [0,0,0]]
print("Before: ")
for row in image:
    print(row)
print("\n")
sr = 0
sc = 0
color = 2
result = solution.floodFill(image, sr, sc, color)
print("After: ")
for row in result:
    print(row)

File: HW/HW11/JavaCode/FloodFill.java
package HW.HW11.JavaCode;

import java.util.LinkedList;
import java.util.Queue;

public class FloodFill{
    public static void main(String[] args) {
        FloodFill solution = new FloodFill();
        int[][] image = {{0,0,0},
                         {0,0,0},
                         };

        // Print Before
        System.out.println("Before:");
        for (int[] row : image) {
            for (int pixel : row) {
                System.out.print(pixel + " ");
            }
            System.out.println();
        }
        System.out.println();
        int sr = 0;
        int sc = 0;
        int color = 0;

        int[][] result = solution.floodFill(image, sr, sc, color);

        // Print the result
        System.out.println("After");
        for (int[] row : result) {
            for (int pixel : row) {
                System.out.print(pixel + " ");
            }
            System.out.println();
        }

    }
    public int[][] floodFill(int[][] image, int sr, int sc, int color){
        int originalColor = image[sr][sc];
        if (originalColor != color) {
            Queue<int[]> queue = new LinkedList<>();
            queue.offer(new int[]{sr, sc});
            image[sr][sc] = color;

            int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            while (!queue.isEmpty()) {
                int[] current = queue.poll();
                for (int[] dir : directions) {
                    int nr = current[0] + dir[0];
                    int nc = current[1] + dir[1];
                    if (nr >= 0 && nr < image.length && nc >= 0 && nc < image[0].length && image[nr][nc] == originalColor) {
                        image[nr][nc] = color;
                        queue.offer(new int[]{nr, nc});
                    }
                }
            }
        }
        return image;
    }
}